src/SkeletalTrapezoidation.cpp

```diff
@@ -0,0 +1,1844 @@
+            { // Snap to start node if it is really close, in order to be able to see 3-way intersection later on more robustly
+                junction = a;
+            }
+            ret.emplace_back(junction, beading->bead_widths[junction_idx], junction_idx);
+        }
+    }
+}
+
+
+std::shared_ptr<SkeletalTrapezoidationJoint::BeadingPropagation> SkeletalTrapezoidation::getOrCreateBeading(node_t* node, ptr_vector_t<BeadingPropagation>& node_beadings)
+{
+    if (! node->data.hasBeading())
+    {
+        if (node->data.bead_count == -1)
+        { // This bug is due to too small marked edges
+            constexpr coord_t nearby_dist = 100; // TODO
+            auto nearest_beading = getNearestBeading(node, nearby_dist);
+            if (nearest_beading)
+            {
+                return nearest_beading;
+            }
+            
+            // Else make a new beading:
+            bool has_marked_edge = false;
+            bool first = true;
+            coord_t dist = std::numeric_limits<coord_t>::max();
+            for (edge_t* edge = node->some_edge; edge && (first || edge != node->some_edge); edge = edge->twin->next)
+            {
+                if (edge->data.isMarked())
+                {
+                    has_marked_edge = true;
+                }
+                assert(edge->to->data.distance_to_boundary >= 0);
+                dist = std::min(dist, edge->to->data.distance_to_boundary + vSize(edge->to->p - edge->from->p));
+                first = false;
+            }
+            RUN_ONCE(logError("Unknown beading for unmarked node!\n"));
+            assert(dist != std::numeric_limits<coord_t>::max());
+            node->data.bead_count = beading_strategy.getOptimalBeadCount(dist * 2);
+        }
+        assert(node->data.bead_count != -1);
+        node_beadings.emplace_back(new BeadingPropagation(beading_strategy.compute(node->data.distance_to_boundary * 2, node->data.bead_count)));
+        node->data.setBeading(node_beadings.back());
+    }
+    assert(node->data.hasBeading());
+    return node->data.getBeading();
+}
+
+std::shared_ptr<SkeletalTrapezoidationJoint::BeadingPropagation> SkeletalTrapezoidation::getNearestBeading(node_t* node, coord_t max_dist)
+{
+    struct DistEdge
+    {
+        edge_t* edge_to;
+        coord_t dist;
+        DistEdge(edge_t* edge_to, coord_t dist)
+        : edge_to(edge_to), dist(dist)
+        {}
+    };
+
+    auto compare = [](const DistEdge& l, const DistEdge& r) -> bool { return l.dist > r.dist; };
+    std::priority_queue<DistEdge, std::vector<DistEdge>, decltype(compare)> further_edges(compare);
+    bool first = true;
+    for (edge_t* outgoing = node->some_edge; outgoing && (first || outgoing != node->some_edge); outgoing = outgoing->twin->next)
+    {
+        further_edges.emplace(outgoing, vSize(outgoing->to->p - outgoing->from->p));
+        first = false;
+    }
+
+    for (coord_t counter = 0; counter < 1000; counter++)
```

Comment:
Magic number. Why 1000 anyway?

Reply:
This counter exists in case there is a long chain of edges without beading. These chains are the outlines of cells between the branches of the skeletal graph, so they shouldn't normally be very long, but they could be. Take the case of a long, curved, thin wall. The skeletal graph will make lines on both ends of the wall, but all vertices in the middle are too shallow to create an edge towards the contour then. Yet the centreline will have many segments. If that's more than 1000 segments it will fail here.

I'll turn this into a macro near the top of the file so that it's easier to edit.